# Final Submission Part 1

In [303]:
import pandas as pd
import matplotlib as mpl
import json
import requests 
from pandas.io.json import json_normalize 
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import matplotlib.pyplot as plt

In [2]:
!conda install -c conda-forge folium --yes
import folium

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    branca-0.3.1               |             py_0          25 KB  conda-forge
    ca-certificates-2019.9.11  |       hecc5488_0         144 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    certifi-2019.9.11          |           py36_0         147 KB  conda-forge
    folium-0.10.0              |             py_0          59 KB  conda-forge
    openssl-1.1.1c             |       h516909a_0         2.1 MB  conda-forge
    altair-3.2.0               |           py36_0         770 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         3.3 MB

The following NEW packages will be INSTAL

In [3]:
!conda install -c anaconda beautifulsoup4

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - beautifulsoup4


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2019.9.11          |           py36_0         154 KB  anaconda
    openssl-1.1.1              |       h7b6447c_0         5.0 MB  anaconda
    beautifulsoup4-4.8.1       |           py36_0         153 KB  anaconda
    ca-certificates-2019.10.16 |                0         131 KB  anaconda
    ------------------------------------------------------------
                                           Total:         5.5 MB

The following packages will be UPDATED:

    beautifulsoup4:  4.7.1-py36_1                     --> 4.8.1-py36_0     anaconda
    ca-certificates: 2019.9.11-hecc5488_0 conda-forge --> 2019.10.16-0     anaconda
    certifi:         2019.9.11-py36_0     conda-forge --> 2019.9.11-py

#### After importing all the important function, I have created three seperate lists with the values for postal code, boroughs, and neighborhoods

In [17]:
import requests
url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
from bs4 import BeautifulSoup as bs

In [18]:
newsoup = bs(url, 'html.parser')
newtable = newsoup.find('table',{'class':'wikitable sortable'})
testset = newtable.find_all('td')
newlist = []
for k in testset:
    print(k.text)
    newlist.append(k.text)
print(len(newlist))

M1A
Not assigned
Not assigned

M2A
Not assigned
Not assigned

M3A
North York
Parkwoods

M4A
North York
Victoria Village

M5A
Downtown Toronto
Harbourfront

M5A
Downtown Toronto
Regent Park

M6A
North York
Lawrence Heights

M6A
North York
Lawrence Manor

M7A
Queen's Park
Not assigned

M8A
Not assigned
Not assigned

M9A
Etobicoke
Islington Avenue

M1B
Scarborough
Rouge

M1B
Scarborough
Malvern

M2B
Not assigned
Not assigned

M3B
North York
Don Mills North

M4B
East York
Woodbine Gardens

M4B
East York
Parkview Hill

M5B
Downtown Toronto
Ryerson

M5B
Downtown Toronto
Garden District

M6B
North York
Glencairn

M7B
Not assigned
Not assigned

M8B
Not assigned
Not assigned

M9B
Etobicoke
Cloverdale

M9B
Etobicoke
Islington

M9B
Etobicoke
Martin Grove

M9B
Etobicoke
Princess Gardens

M9B
Etobicoke
West Deane Park

M1C
Scarborough
Highland Creek

M1C
Scarborough
Rouge Hill

M1C
Scarborough
Port Union

M2C
Not assigned
Not assigned

M3C
North York
Flemingdon Park

M3C
North York
Don Mills South


In [19]:
c = 0
postal_code = []
borough = []
neighborhood = []
for i in newlist:
    postal_code.append(newlist[c])
    c = c+1
    borough.append(newlist[c])
    c = c+1
    neighborhood.append(newlist[c].strip())
    c = c+1
    if c == 864:
        break
    
print(postal_code)
print(borough)
print(neighborhood)

['M1A', 'M2A', 'M3A', 'M4A', 'M5A', 'M5A', 'M6A', 'M6A', 'M7A', 'M8A', 'M9A', 'M1B', 'M1B', 'M2B', 'M3B', 'M4B', 'M4B', 'M5B', 'M5B', 'M6B', 'M7B', 'M8B', 'M9B', 'M9B', 'M9B', 'M9B', 'M9B', 'M1C', 'M1C', 'M1C', 'M2C', 'M3C', 'M3C', 'M4C', 'M5C', 'M6C', 'M7C', 'M8C', 'M9C', 'M9C', 'M9C', 'M9C', 'M1E', 'M1E', 'M1E', 'M2E', 'M3E', 'M4E', 'M5E', 'M6E', 'M7E', 'M8E', 'M9E', 'M1G', 'M2G', 'M3G', 'M4G', 'M5G', 'M6G', 'M7G', 'M8G', 'M9G', 'M1H', 'M2H', 'M3H', 'M3H', 'M3H', 'M4H', 'M5H', 'M5H', 'M5H', 'M6H', 'M6H', 'M7H', 'M8H', 'M9H', 'M1J', 'M2J', 'M2J', 'M2J', 'M3J', 'M3J', 'M4J', 'M5J', 'M5J', 'M5J', 'M6J', 'M6J', 'M7J', 'M8J', 'M9J', 'M1K', 'M1K', 'M1K', 'M2K', 'M3K', 'M3K', 'M4K', 'M4K', 'M5K', 'M5K', 'M6K', 'M6K', 'M6K', 'M7K', 'M8K', 'M9K', 'M1L', 'M1L', 'M1L', 'M2L', 'M2L', 'M3L', 'M4L', 'M4L', 'M5L', 'M5L', 'M6L', 'M6L', 'M6L', 'M7L', 'M8L', 'M9L', 'M1M', 'M1M', 'M1M', 'M2M', 'M2M', 'M3M', 'M4M', 'M5M', 'M5M', 'M6M', 'M6M', 'M6M', 'M6M', 'M7M', 'M8M', 'M9M', 'M9M', 'M1N', 'M1N', 'M2N'

#### Once I have all the needed lists, I create a dataframe to replicate the one on wikipedia

In [20]:
dic = {'PostalCode': postal_code, 'Borough':borough, 'Neighborhood':neighborhood}
df = pd.DataFrame(dic, columns = ['PostalCode', 'Borough', 'Neighborhood'])
df.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

#### I drop the not assigned values (leaving the Queen's Park one) and the replace the Queen's Park neighborhood to the same as borough as asked in the assignment

In [21]:
df.drop( df[ df['Borough'] == "Not assigned" ].index , inplace=True)

In [22]:
df.replace(to_replace = 'Not assigned',value ='Queen\'s Park', inplace = True)

In [23]:
df.head()

PostalCode           Borough      Neighborhood
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront
5        M5A  Downtown Toronto       Regent Park
6        M6A        North York  Lawrence Heights

#### I grouped it by postal code and brorugh since there are multiple neighborhoods for a particular postal code in a borough and then reset index to see the proper shape.

In [24]:
df = df.groupby(['PostalCode','Borough'], sort = False).agg(lambda x: ','.join(x))

In [25]:
df = df.reset_index()

In [26]:
df.shape

(103, 3)

Our Data Frame has 103 rows and 3 columns after some preprocessing

In [27]:
!conda install -c conda-forge geocoder

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geocoder


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2019.9.11          |           py36_0         147 KB  conda-forge
    geocoder-1.38.1            |             py_1          53 KB  conda-forge
    ratelim-0.1.6              |             py_2           6 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         206 KB

The following NEW packages will be INSTALLED:

    geocoder:        1.38.1-py_1      conda-forge
    ratelim:         0.1.6-py_2       conda-forge

The following packages will be UPDATED:

    certifi:         2019.9.11-py36_0 anaconda    --> 2019.9.11-py36_0     conda-forge

The following packages will be DOWNGRADED:

    ca-certificates: 2019.10.1

import geocoder   #import geocoder

#initialize your variable to None  
lat_lng_coords = None  

#loop until you get the coordinates  
while(lat_lng_coords is None):  
    g = geocoder.google('{}, Toronto, Ontario'.format(postal_code[0]))  
    lat_lng_coords = g.latlng  

latitude = lat_lng_coords[0]  
longitude = lat_lng_coords[1]  
### I tried using this code but the loop was taking far too long and so i am going to read the csv file provided

# Part 2

In [32]:
newfile = pd.read_csv('https://cocl.us/Geospatial_data')
newfile

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
..          ...        ...        ...
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [46]:
newfile.rename(columns = {'Postal Code': 'PostalCode'}, inplace = True)
dftest = pd.merge(df, newfile, on = 'PostalCode', how = 'outer')
dftest

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A      Queen's Park   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
..         ...               ...   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                             Harbourfront,Regent Park  43.654260 -79.360636  
3                      Lawrence Heights,Lawrence Manor  43.718518 -79.464763  
4                                         Queen's Park  43.662301 -79.389494  
5                                     Islington Avenue  43.667856 -79.532242  
6                                        Rouge,Malvern  43.806686 -79.194353  
7                                      Don Mills North  43.745906 -79.352188  
8                       Woodbine Gardens,Parkview Hill  43.706397 -79.309937  
9                              Ryerson,Garden District  43.657162 -79.378937  
10                                           Glencairn  43.709577 -79.445073  
11   Cloverdale,Islington,Martin Grove,Princess Gar...  43.650943 -79.554724  
12                Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
13                     Flemingdon Park,Don Mills South  43.725900 -79.340923  
14                                    Woodbine Heights  43.695344 -79.318389  
15                                      St. James Town  43.651494 -79.375418  
16                                  Humewood-Cedarvale  43.693781 -79.428191  
17   Bloordale Gardens,Eringate,Markland Wood,Old B...  43.643515 -79.577201  
18                     Guildwood,Morningside,West Hill  43.763573 -79.188711  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
21                               

# Part 3

#### A lot of the code below follows the new york lab from w3. Please pay attention to the maps and the final few lines of code because that is main. I was having a lot of fun playing around with python loops and codes with the location data, so you might notice some random information lying around, you can ignore that. 

In [42]:
from geopy.geocoders import Nominatim
latitude = 0
while latitude == 0:   
    address = 'Toronto, Ontario, Canada'
    geolocator = Nominatim(user_agent="mynew_app")
    location = geolocator.geocode(address)
    latitude = location.latitude
    longitude = location.longitude
print('The geograpical coordinate of Torronto Canada are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Torronto Canada are 43.653963, -79.387207.


In [43]:
# create map of Manhattan using latitude and longitude values
map_canada = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(dftest['Latitude'], dftest['Longitude'], dftest['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_canada)  
    
map_canada

In [59]:
CLIENT_ID = '1LHIK33PSSXZAVXRQGU05MXT2UT010ITHOATFC5OAAOOSYA0' # your Foursquare ID
CLIENT_SECRET = '0HJM5LOB00UMCYHFE5KUXIHFW1QVXUUBKZOZMILHYYK2OV3W' # your Foursquare Secret
VERSION = '20191029' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 1LHIK33PSSXZAVXRQGU05MXT2UT010ITHOATFC5OAAOOSYA0
CLIENT_SECRET:0HJM5LOB00UMCYHFE5KUXIHFW1QVXUUBKZOZMILHYYK2OV3W


In [52]:
dftest.loc[0, 'Neighborhood']
dftest

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A      Queen's Park   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
..         ...               ...   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                             Harbourfront,Regent Park  43.654260 -79.360636  
3                      Lawrence Heights,Lawrence Manor  43.718518 -79.464763  
4                                         Queen's Park  43.662301 -79.389494  
5                                     Islington Avenue  43.667856 -79.532242  
6                                        Rouge,Malvern  43.806686 -79.194353  
7                                      Don Mills North  43.745906 -79.352188  
8                       Woodbine Gardens,Parkview Hill  43.706397 -79.309937  
9                              Ryerson,Garden District  43.657162 -79.378937  
10                                           Glencairn  43.709577 -79.445073  
11   Cloverdale,Islington,Martin Grove,Princess Gar...  43.650943 -79.554724  
12                Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
13                     Flemingdon Park,Don Mills South  43.725900 -79.340923  
14                                    Woodbine Heights  43.695344 -79.318389  
15                                      St. James Town  43.651494 -79.375418  
16                                  Humewood-Cedarvale  43.693781 -79.428191  
17   Bloordale Gardens,Eringate,Markland Wood,Old B...  43.643515 -79.577201  
18                     Guildwood,Morningside,West Hill  43.763573 -79.188711  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
21                               

In [57]:
c = 0
neigh = []
lat = []
lon = []

for x in dftest.index:    
    neighborhood_latitude = dftest.loc[c, 'Latitude'] # neighborhood latitude value
    neighborhood_longitude = dftest.loc[c, 'Longitude'] # neighborhood longitude value

    neighborhood_name = dftest.loc[c, 'Neighborhood'] # neighborhood name
    neigh.append(neighborhood_name)
    lat.append(neighborhood_latitude)
    lon.append(neighborhood_longitude)
    c=c+1

print(neigh)
print(lat)
print(lon)

['Parkwoods', 'Victoria Village', 'Harbourfront,Regent Park', 'Lawrence Heights,Lawrence Manor', "Queen's Park", 'Islington Avenue', 'Rouge,Malvern', 'Don Mills North', 'Woodbine Gardens,Parkview Hill', 'Ryerson,Garden District', 'Glencairn', 'Cloverdale,Islington,Martin Grove,Princess Gardens,West Deane Park', 'Highland Creek,Rouge Hill,Port Union', 'Flemingdon Park,Don Mills South', 'Woodbine Heights', 'St. James Town', 'Humewood-Cedarvale', 'Bloordale Gardens,Eringate,Markland Wood,Old Burnhamthorpe', 'Guildwood,Morningside,West Hill', 'The Beaches', 'Berczy Park', 'Caledonia-Fairbanks', 'Woburn', 'Leaside', 'Central Bay Street', 'Christie', 'Cedarbrae', 'Hillcrest Village', 'Bathurst Manor,Downsview North,Wilson Heights', 'Thorncliffe Park', 'Adelaide,King,Richmond', 'Dovercourt Village,Dufferin', 'Scarborough Village', 'Fairview,Henry Farm,Oriole', 'Northwood Park,York University', 'East Toronto', 'Harbourfront East,Toronto Islands,Union Station', 'Little Portugal,Trinity', 'East 

In [202]:
urllist = []
n = 0
for x in dftest.index:
    LIMIT = 50
    radius = 300
    newurl = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat[n],
        lon[n],
        radius,
        LIMIT)
    n=n+1
    urllist.append(newurl)
    
urllist

['https://api.foursquare.com/v2/venues/explore?&client_id=1LHIK33PSSXZAVXRQGU05MXT2UT010ITHOATFC5OAAOOSYA0&client_secret=0HJM5LOB00UMCYHFE5KUXIHFW1QVXUUBKZOZMILHYYK2OV3W&v=20191029&ll=43.7532586,-79.3296565&radius=300&limit=50',
 'https://api.foursquare.com/v2/venues/explore?&client_id=1LHIK33PSSXZAVXRQGU05MXT2UT010ITHOATFC5OAAOOSYA0&client_secret=0HJM5LOB00UMCYHFE5KUXIHFW1QVXUUBKZOZMILHYYK2OV3W&v=20191029&ll=43.725882299999995,-79.31557159999998&radius=300&limit=50',
 'https://api.foursquare.com/v2/venues/explore?&client_id=1LHIK33PSSXZAVXRQGU05MXT2UT010ITHOATFC5OAAOOSYA0&client_secret=0HJM5LOB00UMCYHFE5KUXIHFW1QVXUUBKZOZMILHYYK2OV3W&v=20191029&ll=43.6542599,-79.3606359&radius=300&limit=50',
 'https://api.foursquare.com/v2/venues/explore?&client_id=1LHIK33PSSXZAVXRQGU05MXT2UT010ITHOATFC5OAAOOSYA0&client_secret=0HJM5LOB00UMCYHFE5KUXIHFW1QVXUUBKZOZMILHYYK2OV3W&v=20191029&ll=43.718517999999996,-79.46476329999999&radius=300&limit=50',
 'https://api.foursquare.com/v2/venues/explore?&client

In [71]:
len(urllist)

103

In [203]:
count = 0
results = []
for x in urllist:
    results.append(requests.get(urllist[count]).json())
    if count == 102:
        break
    count = count + 1

In [205]:
len(results)
results

[{'meta': {'code': 200, 'requestId': '5db974b3787dba003355cb18'},
  'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
   'headerLocation': 'Parkwoods - Donalda',
   'headerFullLocation': 'Parkwoods - Donalda, Toronto',
   'headerLocationGranularity': 'neighborhood',
   'totalResults': 1,
   'suggestedBounds': {'ne': {'lat': 43.755958602700005,
     'lng': -79.32592554390956},
    'sw': {'lat': 43.7505585973, 'lng': -79.33338745609043}},
   'groups': [{'type': 'Recommended Places',
     'name': 'recommended',
     'items': [{'reasons': {'count': 0,
        'items': [{'summary': 'This spot is popular',
          'type': 'general',
          'reasonName': 'globalInteractionReason'}]},
       'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
        'name': 'Brookbanks Park',
        'location': {'address': 'Toronto',
         'lat': 43.751976046055574,
         'lng': -79.33214044722958,
      

In [206]:
helplist = []
c = 0
for i in results:
    if results[c]['response']['totalResults'] == 0:
        print(results[c]['response']['headerLocation'], 'has no places')
        c = c+1
    else:
        helplist.append(results[c]['response']['groups'][0]['items'])
        c = c+1
    
    
len(helplist)

Edenbridge - Humber Valley has no places
Malvern has no places
Islington - City Centre West has no places
Toronto has no places
Toronto has no places
Clanton Park has no places
Clairlea - Birchmount has no places
St. Andrew - Windfields has no places
Willowdale has no places
Toronto has no places
Toronto has no places
High Park has no places
Wexford has no places
Scarborough has no places
Toronto has no places
Rexdale has no places
Toronto has no places
Rouge has no places


85

In [191]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [198]:
newFrame = pd.DataFrame(columns = ['name', 'categories', 'lat', 'lng'])

In [199]:
c=0
while c != 85:
    filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
    nearby_venues = json_normalize(helplist[c])
    nearby_venues = nearby_venues.loc[:, filtered_columns]
    # filter the category for each row
    nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)
    # clean columns
    nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]
    c = c+1
    newFrame = newFrame.append(nearby_venues, ignore_index = True)

In [200]:
newFrame

name  \
0                                    Brookbanks Park   
1                             Victoria Village Arena   
2                                        Tim Hortons   
3                                          Portugril   
4                                           The Frig   
5          Eglinton Ave E & Sloane Ave/Bermondsey Rd   
6                                         Pizza Nova   
7                                         Cash Money   
8                                   Roselle Desserts   
9                                      Tandem Coffee   
10   Toronto Cooper Koo Family Cherry St YMCA Centre   
11                               Body Blitz Spa East   
12                                Morning Glory Cafe   
13                                The Extension Room   
14                           Buster's Sea Cove Truck   
15                          Enoch Turner Schoolhouse   
16                                        Percy Park   
17                                             Roots   
18             Kitchen Stuff Plus (Clearance Outlet)   
19                                       Tim Hortons   
20                               Ardene Shoes Outlet   
21                    Rinx Real Entertainment Centre   
22                                    Urban Behavior   
23                                Kitchen Stuff Plus   
24                                         Bluenotes   
25                           International Clothiers   
26                                      Urban Planet   
27                                      Queen's Park   
28                                          Mercatto   
29                                           Nando's   
..                                               ...   
450                                            Maman   
451                                        King Taps   
452                      Sam James Coffee Bar (SJCB)   
453                         The Keg Steakhouse & Bar   
454                                Pizzeria Libretto   
455                    Lambton Kingsway Outdoor Pool   
456                                 Storm Crow Manor   
457                                DanceLifeX Centre   
458                                        The Alley   
459                                            Smith   
460                               Sansotei Ramen 三草亭   
461                                         Fabarnak   
462                                     Como En Casa   
463                            Ho's Team Barber Shop   
464                                   Wickson Social   
465                                Glad Day Bookshop   
466                            Rorschach Brewing Co.   
467                       Leslieville Farmers Market   
468                                      Chick-n-Joy   
469                           Amin Car Repair Garage   
470                             The Ashbridge Estate   
471                             TTC Russell Division   
472                          Jonathan Ashbridge Park   
473                                         Eggsmart   
474                                        Dollarama   
475                                       Wingporium   
476                                   Healthy Planet   
477                                         7-Eleven   
478                                             Rona   
479                                       McDonald's   

                     categories        lat        lng  
0                          Park  43.751976 -79.332140  
1                  Hockey Arena  43.723481 -79.315635  
2                   Coffee Shop  43.725517 -79.313103  
3         Portuguese Restaurant  43.725819 -79.312785  
4             French Restaurant  43.727051 -79.317418  
5                  Intersection  43.726086 -79.313620  
6                   Pizza Place  43.725824 -79.312860  
7    Financial or Legal Service  43.725486 -79.312665  
8                        Bakery  43.653447 -79.362017  
9              

In [201]:
finalFrame = newFrame

In [207]:
def getNearbyVenues(names, latitudes, longitudes, radius=300):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()['response']['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [208]:
canada_venues = getNearbyVenues(names=dftest['Neighborhood'],
                                   latitudes=dftest['Latitude'],
                                   longitudes=dftest['Longitude']
                                  )


Parkwoods
Victoria Village
Harbourfront,Regent Park
Lawrence Heights,Lawrence Manor
Queen's Park
Islington Avenue
Rouge,Malvern
Don Mills North
Woodbine Gardens,Parkview Hill
Ryerson,Garden District
Glencairn
Cloverdale,Islington,Martin Grove,Princess Gardens,West Deane Park
Highland Creek,Rouge Hill,Port Union
Flemingdon Park,Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Bloordale Gardens,Eringate,Markland Wood,Old Burnhamthorpe
Guildwood,Morningside,West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor,Downsview North,Wilson Heights
Thorncliffe Park
Adelaide,King,Richmond
Dovercourt Village,Dufferin
Scarborough Village
Fairview,Henry Farm,Oriole
Northwood Park,York University
East Toronto
Harbourfront East,Toronto Islands,Union Station
Little Portugal,Trinity
East Birchmount Park,Ionview,Kennedy Park
Bayview Village
CFB Toronto,Downsview East
The Danforth West,Riverdale
Design E

In [209]:
canada_venues.shape

(1070, 7)

In [210]:
print('There are {} uniques categories.'.format(len(canada_venues['Venue Category'].unique())))

There are 218 uniques categories.


In [211]:
# one hot encoding
canada_onehot = pd.get_dummies(canada_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
canada_onehot['Neighborhood'] = canada_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [canada_onehot.columns[-1]] + list(canada_onehot.columns[:-1])
canada_onehot = canada_onehot[fixed_columns]

canada_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Airport Food Court  \
0            0                  0               0                   0   
1            0                  0               0                   0   
2            0                  0               0                   0   
3            0                  0               0                   0   
4            0                  0               0                   0   

   Airport Gate  Airport Lounge  Airport Terminal  American Restaurant  \
0             0               0                 0                    0   
1             0               0                 0                    0   
2             0               0                 0                    0   
3             0               0                 0                    0   
4             0               0                 0                    0   

   Art Gallery  Arts & Crafts Store  ...  Trail  \
0            0                    0  ...      0   
1            0                    0  ...      0   
2            0                    0  ...      0   
3            0                    0  ...      0   
4            0                    0  ...      0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                      0                0         0          0            0   
1                      0                0         0          0            0   
2                      0                0         0          0            0   
3                      0                0         0          0            0   
4                      0                0         0          0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 218 columns]

In [212]:
canada_grouped = canada_onehot.groupby('Neighborhood').mean().reset_index()
canada_grouped

Neighborhood  Yoga Studio  \
0                              Adelaide,King,Richmond     0.000000   
1                                           Agincourt     0.000000   
2                               Alderwood,Long Branch     0.000000   
3       Bathurst Manor,Downsview North,Wilson Heights     0.000000   
4                                     Bayview Village     0.000000   
5                    Bedford Park,Lawrence Manor East     0.000000   
6                                         Berczy Park     0.000000   
7                          Birch Cliff,Cliffside West     0.000000   
8   Bloordale Gardens,Eringate,Markland Wood,Old B...     0.000000   
9          Brockton,Exhibition Place,Parkdale Village     0.000000   
10  Business Reply Mail Processing Centre 969 Eastern     0.000000   
11  CN Tower,Bathurst Quay,Island airport,Harbourf...     0.000000   
12                         Cabbagetown,St. James Town     0.000000   
13                                Caledonia-Fairbanks     0.000000   
14              Canada Post Gateway Processing Centre     0.000000   
15                                          Cedarbrae     0.000000   
16                                 Central Bay Street     0.000000   
17            Chinatown,Grange Park,Kensington Market     0.000000   
18                                           Christie     0.000000   
19                               Church and Wellesley     0.000000   
20              Clarks Corners,Sullivan,Tam O'Shanter     0.000000   
21      Cliffcrest,Cliffside,Scarborough Village West     0.000000   
22                      Commerce Court,Victoria Hotel     0.000000   
23                                         Davisville     0.000000   
24                                   Davisville North     0.000000   
25  Deer Park,Forest Hill SE,Rathnelly,South Hill,...     0.000000   
26        Del Ray,Keelesdale,Mount Dennis,Silverthorn     0.000000   
27            Design Exchange,Toronto Dominion Centre     0.000000   
28                                    Don Mills North     0.000000   
29  Dorset Park,Scarborough Town Centre,Wexford He...     0.000000   
..                                                ...          ...   
55                                            Leaside     0.000000   
56                            Little Portugal,Trinity     0.025000   
57                         Moore Park,Summerhill East     0.000000   
58                                 North Toronto West     0.000000   
59                              Parkdale,Roncesvalles     0.000000   
60                                          Parkwoods     0.000000   
61                                       Queen's Park     0.000000   
62                                           Rosedale     0.000000   
63                                           Roselawn     0.000000   
64                                  Runnymede,Swansea     0.000000   
65                            Ryerson,Garden District     0.000000   
66                                Scarborough Village     0.000000   
67                                     St. James Town     0.000000   
68                    Stn A PO Boxes 25 The Esplanade     0.000000   
69                                    Studio District     0.000000   
70                  The Annex,North Midtown,Yorkville     0.000000   
71                                        The Beaches     0.000000   
72                      The Beaches West,India Bazaar     0.000000   
73                        The Danforth West,Riverdale     0.045455   
74        The Kingsway,Montgomery Road,Old Mill North     0.000000   
75                                   Thorncliffe Park     0.066667   
76                                   Victoria Village     0.000000   
77                                          Westmount     0.000000   
78                                             Weston     0.000000   
79                                   Willowdale South     0.000000   
80                                    Willowdale We

In [213]:
canada_grouped.shape

(85, 218)

In [251]:
num_top_venues = 5

for hood in canada_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = canada_grouped[canada_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
                 venue  freq
0           Steakhouse  0.08
1          Coffee Shop  0.06
2  American Restaurant  0.06
3                  Bar  0.06
4     Asian Restaurant  0.06


----Agincourt----
                 venue  freq
0       Breakfast Spot   0.5
1       Sandwich Place   0.5
2  Monument / Landmark   0.0
3   Mac & Cheese Joint   0.0
4     Malay Restaurant   0.0


----Alderwood,Long Branch----
         venue  freq
0     Pharmacy   0.2
1          Gym   0.2
2  Pizza Place   0.2
3  Coffee Shop   0.2
4          Pub   0.2


----Bathurst Manor,Downsview North,Wilson Heights----
              venue  freq
0       Coffee Shop  0.15
1       Pizza Place  0.08
2    Sandwich Place  0.08
3        Restaurant  0.08
4  Sushi Restaurant  0.08


----Bayview Village----
                 venue  freq
0       Shopping Plaza   1.0
1          Yoga Studio   0.0
2  Monument / Landmark   0.0
3   Mac & Cheese Joint   0.0
4     Malay Restaurant   0.0


----Bedford Park,Lawrence Man

In [306]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [307]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = canada_grouped['Neighborhood']

for ind in np.arange(canada_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(canada_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                         Adelaide,King,Richmond            Steakhouse   
1                                      Agincourt        Breakfast Spot   
2                          Alderwood,Long Branch                   Gym   
3  Bathurst Manor,Downsview North,Wilson Heights           Coffee Shop   
4                                Bayview Village        Shopping Plaza   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0           Coffee Shop                   Bar        American Restaurant   
1        Sandwich Place         Women's Store             Discount Store   
2           Coffee Shop           Pizza Place                   Pharmacy   
3   Fried Chicken Joint           Pizza Place  Middle Eastern Restaurant   
4         Women's Store                 Diner       Fast Food Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0      Asian Restaurant   Japanese Restaurant                 Hotel   
1  Fast Food Restaurant        Farmers Market    Falafel Restaurant   
2                   Pub        Cosmetics Shop    Dim Sum Restaurant   
3         Deli / Bodega            Restaurant        Sandwich Place   
4        Farmers Market    Falafel Restaurant           Event Space   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Seafood Restaurant      Sushi Restaurant                   Café  
1           Event Space  Ethiopian Restaurant      Electronics Store  
2    Falafel Restaurant           Event Space   Ethiopian Restaurant  
3         Shopping Mall           Supermarket       Sushi Restaurant  
4  Ethiopian Restaurant     Electronics Store    Dumpling Restaurant

In [308]:
# set number of clusters
kclusters = 5 

canada_grouped_clustering = canada_grouped.drop('Neighborhood', 1)


# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(canada_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 4, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [309]:

# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

canada_merged = dftest

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
canada_merged = canada_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood', how = 'right')

canada_merged.head() # check the last columns!

PostalCode           Borough                     Neighborhood   Latitude  \
0        M3A        North York                        Parkwoods  43.753259   
1        M4A        North York                 Victoria Village  43.725882   
2        M5A  Downtown Toronto         Harbourfront,Regent Park  43.654260   
3        M6A        North York  Lawrence Heights,Lawrence Manor  43.718518   
4        M7A      Queen's Park                     Queen's Park  43.662301   

   Longitude  Cluster Labels       1st Most Common Venue  \
0 -79.329656               2                        Park   
1 -79.315572               1  Financial or Legal Service   
2 -79.360636               1        Gym / Fitness Center   
3 -79.464763               1              Clothing Store   
4 -79.389494               1                 Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Diner                 Field  Fast Food Restaurant   
1             Coffee Shop          Hockey Arena           Pizza Place   
2             Coffee Shop                   Spa        Breakfast Spot   
3  Furniture / Home Store         Women's Store           Coffee Shop   
4         Bubble Tea Shop    Italian Restaurant                  Lake   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0         Farmers Market    Falafel Restaurant           Event Space   
1  Portuguese Restaurant     French Restaurant          Intersection   
2             Food Truck                Bakery                  Park   
3            Event Space              Boutique     Accessories Store   
4         Sandwich Place      Sushi Restaurant                  Café   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0   Ethiopian Restaurant     Electronics Store    Dumpling Restaurant  
1                  Diner    Falafel Restaurant            Event Space  
2         History Museum        Farmers Market     Falafel Restaurant  
3           Gourmet Shop    Dim Sum Restaurant         Farmers Market  
4  Portuguese Restaurant                  Park       Cuban Restaurant

In [311]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(canada_merged['Latitude'], canada_merged['Longitude'], canada_merged['Neighborhood'], canada_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [312]:
canada_merged.loc[canada_merged['Cluster Labels'] == 0, canada_merged.columns[[1] + list(range(5, canada_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
16         York               0                 Field                   Gym   
32  Scarborough               0            Playground         Women's Store   
38  Scarborough               0     Convenience Store            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
16            Playground               Dog Run    Dim Sum Restaurant   
32    Dim Sum Restaurant  Fast Food Restaurant        Farmers Market   
38         Women's Store                 Diner  Fast Food Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
16  Fast Food Restaurant        Farmers Market    Falafel Restaurant   
32    Falafel Restaurant           Event Space  Ethiopian Restaurant   
38        Farmers Market    Falafel Restaurant           Event Space   

   9th Most Common Venue 10th Most Common Venue  
16           Event Space   Ethiopian Restaurant  
32     Electronics Store    Dumpling Restaurant  
38  Ethiopian Restaurant      Electronics Store

In [313]:
canada_merged.loc[canada_merged['Cluster Labels'] == 1, canada_merged.columns[[1] + list(range(5, canada_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
1          North York               1  Financial or Legal Service   
2    Downtown Toronto               1        Gym / Fitness Center   
3          North York               1              Clothing Store   
4        Queen's Park               1                 Coffee Shop   
7          North York               1                Tennis Court   
8           East York               1              Breakfast Spot   
9    Downtown Toronto               1                 Coffee Shop   
10         North York               1         Japanese Restaurant   
13         North York               1            Asian Restaurant   
14          East York               1                  Beer Store   
15   Downtown Toronto               1                 Coffee Shop   
17          Etobicoke               1                    Pharmacy   
18        Scarborough               1           Electronics Store   
19       East Toronto               1        Other Great Outdoors   
20   Downtown Toronto               1                Concert Hall   
21               York               1               Women's Store   
22        Scarborough               1           Korean Restaurant   
23          East York               1         Sporting Goods Shop   
24   Downtown Toronto               1                 Coffee Shop   
25   Downtown Toronto               1               Grocery Store   
26        Scarborough               1        Caribbean Restaurant   
27         North York               1                 Golf Course   
28         North York               1                 Coffee Shop   
29          East York               1           Indian Restaurant   
30   Downtown Toronto               1                  Steakhouse   
31       West Toronto               1        Gym / Fitness Center   
33         North York               1              Clothing Store   
36   Downtown Toronto               1                 Coffee Shop   
37       West Toronto               1                         Bar   
39         North York               1              Shopping Plaza   
..                ...             ...                         ...   
60         North York               1                Liquor Store   
61    Central Toronto               1        Gym / Fitness Center   
62    Central Toronto               1     Health & Beauty Service   
65        Scarborough               1           Indian Restaurant   
66         North York               1                        Park   
67    Central Toronto               1                         Gym   
68    Central Toronto               1                    Pharmacy   
70          Etobicoke               1                 Coffee Shop   
73    Central Toronto               1                    Boutique   
74    Central Toronto               1              Sandwich Place   
75       West Toronto               1                Tennis Court   
76        Mississauga               1                 Coffee Shop   
79    Central Toronto               1                Dessert Shop   
80   Downtown Toronto               1                        Café   
81       West Toronto               1            Sushi Restaurant   
82        Scarborough               1        Fast Food Restaurant   
83    Central Toronto               1                        Park   
84   Downtown Toronto               1                        Café   
86    Central Toronto               1                 Coffee Shop   
87   Downtown Toronto               1                 Coffee Shop   
90        Scarborough               1        Fast Food Restaurant   
91   Downtown Toronto               1                  Campground   
92   Downtown Toronto               1                Cocktail Bar   
93          Etobicoke               1                         Gym   
96   Downtown Toronto               1                  Restaurant   
97   Downtown Toronto               1                 Coffee Shop   
98          

# Final Assessment  
after trying to find the perfect k to fit the cada data, the graph showed no elbow point except for a slight shift at k =5. After looking at the data above, one can say that the clustering method used was not the best since cluster 1 is taking the majority of the places available. This could also mean that there is not much commonality between what canadians from different parts of torronto do. I say that this is also the case because there were multiple places with very few location data available making it hard to identify what the most done activties are.